### PR TITLE
[Relay][Quantization] Extend FakeQuantizationToInteger to more ops

### PR DESCRIPTION
**Pull request by mbrookhart**
Adding more ops to support QAT BERT. I also refactored the tests for easier extensions.

I marked this as WIP because the result isn't matching tflite, I need to do some more comparisons over the next day or two  to isolate the problem. 

I'm opening the PR now because supporting ops with multiple outputs required refactoring how I handle quantization specific types. Bringing it into the make ir namespace allows me to do what I need to do, but it also opens up a question of where or not this could be more used more generally in other places for quantization?

cc @jroesch @masahi @anijain2305 @jwfromm

### CODE REVIEW

**Comment by masahi** [comment]
> I marked this as WIP because the result isn't matching tflite

How big is the difference? We should evaluate some accuracy metric on a whole dataset.

**Comment by mbrookhart** [comment]
I've only tested random data so far, looking for a dataset I can test with tomorrow

**Comment by mbrookhart** [comment]
Sorry for the delay on this. I've hit some very subtle accuracy bugs on a BERT model I was using that made me doubt this. I see slightly differences (1.5% rmse) on all comparisons on that model (tflite->onnxruntime, tflite->tvm, tflite->tvm + this, tvm->onnx, etc) I think that might a complication to that model in particular.

File: python/tvm/relay/transform/fake_quantization_to_integer.py [hunk 1]
```diff
@@ -102,7 +104,7 @@ def bias_add(expr, type_map):
             out_dtype=xt.dtype,
```
[the remark this answers is not in the document]
Comment (mbrookhart) [reply python/tvm/relay/transform/fake_quantization_to_integer.py]: Thanks for finding this! I've fixed it and added a test case that hits it.

**Comment by mbrookhart** [comment]
@jwfromm @masahi @jroesch @anijain2305 can you take a look?

**Review by elvin-n — APPROVED** [review]

File: python/tvm/relay/transform/fake_quantization_to_integer.py [hunk 2]
```diff
@@ -163,4 +196,128 @@ def clip(expr, type_map):
...
+
+@register_fake_quantization_to_integer("nn.pad")
+def pad(expr, type_map):
+    """Rewite an nn.pad op"""
+    arg = expr.args[0]
+    t = type_map[arg]
+    pad_value = expr.args[1]
+    if pad_value in type_map:
+        pad_t = type_map[pad_value]
+        if not tvm.ir.structural_equal(t, pad_t):
```
Comment (jwfromm) [inline python/tvm/relay/transform/fake_quantization_to_integer.py:210]: a comment explaining the logic here would be helpful.

File: src/relay/transforms/fake_quantization_to_integer.cc [hunk 3]
```diff
@@ -148,7 +93,8 @@ class SubgraphExtractor : public ExprVisitor {
   const AffineTypeMap GetAffineTypes() { return affine_types_; }
   void VisitExpr(const Expr& expr) override {
     if (expr.as<CallNode>() == nullptr && expr.as<OpNode>() == nullptr &&
```
Comment (jwfromm) [inline src/relay/transforms/fake_quantization_to_integer.cc:95]: can you add a comment explaining this check and why it implies `fake_quantization = false`? It's a little confusing to me.

File: src/relay/transforms/fake_quantization_to_integer.cc [hunk 4]
```diff
@@ -226,18 +173,37 @@ class SubgraphMutator : public ExprMutator {
       // Call the rewrite
       Array<ObjectRef> vals = fqfq[op](expr, affine_types_);
       // Save teh outputs of the rewrite
```
Comment (jwfromm) [inline src/relay/transforms/fake_quantization_to_integer.cc:175]: might as well fix this typo while adding comments.

**Review by jwfromm — APPROVED** [review]
A few nitpicks on comments but this overall looks excellent. The one thing that seems missing is a brief tutorial that explains the concept of fake quantization and shows how to import and run a fake quantized model. We can differ writing that tutorial for a later PR but we should try not to drop it.

**Comment by mbrookhart** [comment]
@elvin-n @jwfromm @masahi @anijain2305 This is finally green after the CI chaos last week, anyone want to take another look before merging?

**Review by masahi — APPROVED** [review]

**Comment by masahi** [comment]
Thanks @mbrookhart @jwfromm @elvin-n

**Pull request merged**